File: GOALx/src/component/component.go
package component

import (
	"html/template"
	"strings"
)

type Component struct {
	name     string
	template template.Template
	style    string
	script   string
}

func (c *Component) Render(props map[string]any) string {
	builder := new(strings.Builder)
	c.template.Execute(builder, props)
	return builder.String()
}

File: Express/components/base/component.js
import * as cheerio from "cheerio";

export default function Component(name, htmlFn, style, script) {
  function render(props) {
    const $ = cheerio.load(htmlFn(props));

    if (style) {
      $.root().append(/* html */ `<div 
            hx-trigger="revealed once" 
            hx-get="/api/style/${name}.css" 
            hx-swap="beforeend" 
            hx-target="style[hx-style]"
            ></div>`);
    }

    return $.html();
  }

  return {
    render,
    style,
    name,
  };
}
